File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    answer = 0
    queue = deque([(begin, 0)])
    n = len(words[0])
    if not target in words: return 0

    def check(str1, str2, n):
        count = 0
        for i in range(n):
            if str1[i] != str2[i]: 
                count += 1
        if count == 1: 
            return True
        else: return False
            
    while queue:
        current_str, count = queue.popleft()
        if current_str == target: 
            return count
        for word in words:
            if word == current_str: continue
            if check(word, current_str, n): 
                queue.append((word, count+1))
                 
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        
        Queue<Node> queue = new ArrayDeque();
        queue.add(new Node(begin, 0));
        
        while (!queue.isEmpty()) {
            
            Node targetNode = queue.poll();
            System.out.println(targetNode.word + " " + targetNode.count);
                        
            // target과 동일한 단어일 경우 return
            if (targetNode.word.equals(target)) {
                return targetNode.count;
            }
            
            // words에 바꿀 수 있는 단어 있나 확인
            for (String word: words) {
                if (canChange(word, targetNode.word)) {
                    queue.add(new Node(word, targetNode.count + 1));
                }
            }

            if (targetNode.count > words.length) break;
        }
        return answer;
    }
    
    // 한글자만 다른지 확인
    public boolean canChange(String word1, String word2) {
        int cnt = 0;
        for (int i = 0; i < word1.length(); i++) {
            if (word1.charAt(i) != word2.charAt(i)) cnt++;
        }
        return cnt == 1;
    }
}



class Node {
    String word;
    int count;
    
    Node(String word, int count) {
        this.word = word;
        this.count = count;
    }
}